### ISSUE TITLE
Argon2 - Freeing memory after use

### ISSUE THREAD

**Issue by lfpraca** [RustCrypto/password-hashes#478]
When using argon2, every time a password is verified, the memory used by the crate increases little by little until stabilizing at around 1.5GiB on my local machine (I am using the default configuration of 19MiB for argon2id) and remains allocated even after the verification is completed and the verifier is dropped. But most of the time it is more valuable to me if argon2 has less memory pre-allocated even if it takes a little more time to verify passwords. Is there a way to configure the verifier (`Argon2::default().verify_password`) to free the memory after verifying a password? I couldn't find information regarding this in the documentation.

Thank you

**Comment by tarcieri** [comment]
Can you post a minimal self-contained repro?

The memory buffer is just a `Vec` which is freed on `Drop` at the end of this 2-line function: https://github.com/RustCrypto/password-hashes/blob/master/argon2/src/lib.rs#L230

Aside from that, there is nothing but stack allocations. The whole crate is heapless/`no_std` friendly. With SIMD disabled there is no unsafe code, even.

Are you sure you're not hanging on to memory elsewhere, unrelated to this crate?

**Comment by tarcieri** [comment]
I see it use 1GB of memory then drop down to 20MB of memory. Granted, this is on macOS.

Are you sure that isn't just memory which has been returned to the heap and hasn't been returned from the heap to the kernel?

**Comment by lfpraca** [comment]
I did a test turning the number down from 300 to 3 verifications per run and running them 80 at a time, I waited for the 80 that I ran to complete before starting a new batch of 80. I could eventually make my computer (running linux) run out of memory and crash, which makes me think this is not what is happening, but I could be wrong

**Comment by lfpraca** [comment]
I tried changing the program to allocate memory with something else, I used a large vector. With it the memory went up to ~180MiB then down at the time "Freed" is printed. In the end the program was using the same amount of memory as when it started

[main.rs.txt](https://github.com/RustCrypto/password-hashes/files/13468487/main.rs.txt)

**Comment by tarcieri** [comment]
There is literally nowhere else in `argon2` crate that touches dynamic memory than this: https://github.com/RustCrypto/password-hashes/blob/d25ea61/argon2/src/lib.rs#L230

I simply don't see any room for this crate to be at fault. It's just allocating a `vec!` and immediately dropping it when it's done.

Perhaps report this upstream to https://github.com/rust-lang/rust ? Otherwise I don't see anywhere a bug could possibly exist in this crate.

**Comment by tarcieri** [comment]
I tried running this on Debian 12 in a Docker container with a Linux 5.15 kernel and glibc 2.36.

RSS as reported by `ps` seemed fairly constant at around ~8000, going up to ~9000 then back down to ~7000 and repeating. This is with your original example.

I'm unable to reproduce this at all. It seems like something with your environment?

**Comment by tarcieri** [comment]
Closing as unreproducible (and even if it were, I don't see anything actionable here on this crate's part)

**Comment by mrjackwills** [comment]
I'm not sure if the correct procedure is to comment here, or to create a new issue, but I am having the same issue and am able to replicate it, on both my home laptop and a remote server.

The code is as mentioned aboce;
```rust
use argon2::{
    password_hash::{PasswordHash, PasswordVerifier},
    Argon2,
};

fn main() {
    std::thread::sleep(std::time::Duration::from_secs(5));
	println!("Starting");
        for _ in 1..=30 {
            let parsed_hash = PasswordHash::new(
                r#"$argon2i$v=19$m=19000,t=2,p=1$a280Y05CQjVrWWxZQjBEaw$moMWFS2AVx0lvSndhzFrWQ"#,
            )
            .unwrap();
            let _ = Argon2::default().verify_password("password".as_bytes(), &parsed_hash);
        }
    println!("Finished");
    std::thread::sleep(std::time::Duration::from_secs(10_000));
}
```

After "Finished" is printed to stdout, the application is still reportedly *holding* on to 150mb of memory.


I don't know what the most reliable way to record memory usage/held memory is. I was was just a combination of `htop`, `ps`, and `top`. I can provide some kind of screen recording would offer any kind of help.

**Comment by cerpins** [comment]
I had a similar issue but realized it was silly. 
Argon for me tops out at around 100MiB of RAM. The issue is that this was duplicated across many workers in my webserver, which by default launches >15 workers (1.5GiB of RAM used!) and crashes my tiny server. [For example, see worker count for actix web.](https://docs.rs/actix-web/4.11.0/actix_web/struct.HttpServer.html#method.bind)

It seems to me an easy mistake to overlook, so I'm leaving a comment if someone runs into it.
